File: node-dock/app/dock-llm/src/test/kotlin/dev/orbit/dock/llm/SafeComputeTest.kt
package dev.orbit.dock.llm

import com.google.common.truth.Truth.assertThat
import kotlin.random.Random
import kotlin.test.Test

/**
 * The compute tool's evaluator — the dock's bounded "run code" stand-in. Must do
 * the arithmetic/random/branch a model needs ("pick a number, if >5 say hi") and
 * REJECT anything resembling general code (names, calls, statements).
 */
class SafeComputeTest {
    private val c = SafeCompute(Random(42))   // seeded → deterministic random

    @Test fun arithmeticWithPrecedence() {
        assertThat(c.eval("3 + 4 * 2")).isEqualTo("11")
        assertThat(c.eval("(3 + 4) * 2")).isEqualTo("14")
        assertThat(c.eval("10 / 4")).isEqualTo("2.5")
        assertThat(c.eval("10 % 3")).isEqualTo("1")
        assertThat(c.eval("-5 + 2")).isEqualTo("-3")
    }

    @Test fun comparisonsReturnBool() {
        assertThat(c.eval("7 > 5")).isEqualTo("true")
        assertThat(c.eval("3 >= 5")).isEqualTo("false")
        assertThat(c.eval("4 == 4")).isEqualTo("true")
        assertThat(c.eval("4 != 4")).isEqualTo("false")
    }

    @Test fun randomIsInRangeInclusive() {
        repeat(50) {
            val v = SafeCompute(Random(it.toLong())).eval("random(1,10)").toInt()
            assertThat(v).isAtLeast(1)
            assertThat(v).isAtMost(10)
        }
    }

    @Test fun randomBranchComparison() {
        // "random(1,10) > 5" → a boolean the model branches on.
        val r = c.eval("random(1,10) > 5")
        assertThat(r).isAnyOf("true", "false")
    }

    @Test fun rejectsGeneralCode() {
        assertThat(c.eval("import random")).startsWith("error")
        assertThat(c.eval("print('hi')")).startsWith("error")
        assertThat(c.eval("x = 5")).startsWith("error")
        assertThat(c.eval("os.system('rm')")).startsWith("error")
        assertThat(c.eval("")).startsWith("error")
    }

    @Test fun rejectsMalformed() {
        assertThat(c.eval("3 +")).startsWith("error")
        assertThat(c.eval("random(5,1)")).startsWith("error")   // a>b
        assertThat(c.eval("(3 + 4")).startsWith("error")        // unbalanced
    }
}

File: orbit-station/server/src/modules/brain/safe-compute.test.ts
/**
 * Ported from the dock's SafeComputeTest.kt — the compute tool's evaluator
 * must do the arithmetic/random/branch a model needs and REJECT anything
 * resembling general code (names, calls, statements).
 */
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { SafeCompute } from './safe-compute.js';

/** deterministic rng for repeatable vectors (mulberry32). */
function seeded(seed: number): () => number {
  let a = seed >>> 0;
  return () => {
    a = (a + 0x6d2b79f5) >>> 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

const c = new SafeCompute(seeded(42));

test('arithmetic with precedence', () => {
  assert.equal(c.eval('3 + 4 * 2'), '11');
  assert.equal(c.eval('(3 + 4) * 2'), '14');
  assert.equal(c.eval('10 / 4'), '2.5');
  assert.equal(c.eval('10 % 3'), '1');
  assert.equal(c.eval('-5 + 2'), '-3');
});

test('comparisons return bool', () => {
  assert.equal(c.eval('7 > 5'), 'true');
  assert.equal(c.eval('3 >= 5'), 'false');
  assert.equal(c.eval('4 == 4'), 'true');
  assert.equal(c.eval('4 != 4'), 'false');
});

test('random is in range inclusive', () => {
  for (let seed = 0; seed < 50; seed++) {
    const v = Number(new SafeCompute(seeded(seed)).eval('random(1,10)'));
    assert.ok(v >= 1, `seed ${seed}: ${v} >= 1`);
    assert.ok(v <= 10, `seed ${seed}: ${v} <= 10`);
  }
});

test('random branch comparison', () => {
  const r = c.eval('random(1,10) > 5');
  assert.ok(r === 'true' || r === 'false', r);
});

test('rejects general code', () => {
  assert.match(c.eval('import random'), /^error/);
  assert.match(c.eval("print('hi')"), /^error/);
  assert.match(c.eval('x = 5'), /^error/);
  assert.match(c.eval("os.system('rm')"), /^error/);
  assert.match(c.eval(''), /^error/);
});

test('rejects malformed', () => {
  assert.match(c.eval('3 +'), /^error/);
  assert.match(c.eval('random(5,1)'), /^error/);   // a>b
  assert.match(c.eval('(3 + 4'), /^error/);        // unbalanced
});
